### ISSUE TITLE
The language server crashed, Linux and Apparent Null Pointer

### ISSUE THREAD

**Issue by tomatghub** [microsoft/vscode-cpptools#3509]
**Type: LanguageService**

**Describe the bug**
- OS and Version: Linux version 5.0.9-arch1-1-ARCH
- VS Code Version: 1.33.1; 51b0b28134d51361cf996d2f0a1c698247aeabd8; x64
- C/C++ Extension Version: ms-vscode.cpptools-0.22.1
- Other extensions you installed (and if the issue persists after disabling them): Python. Didn't disable.
- A clear and concise description of what the bug is.

This problem shows up the same way all the other server crashes do:

**To Reproduce**
Hover over a definition in a C source file:

```
[Error - 4:48:26 PM] Connection to server got closed. Server will not be restarted.
[Error - 4:48:26 PM] Request textDocument/hover failed.
Error: Connection got disposed.
	at Object.dispose (/home/xxxxxx/.vscode-oss/extensions/ms-vscode.cpptools-0.22.1/node_modules/vscode-jsonrpc/lib/main.js:876:25)
	at Object.dispose (/home/xxxxxx/.vscode-oss/extensions/ms-vscode.cpptools-0.22.1/node_modules/vscode-languageclient/lib/client.js:71:35)
And so on...
```
Attaching gdb to the server process and then hovering gives:

```
Thread 19 "Microsoft.VSCod" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fa8cbfff700 (LWP 14031)]
0x0000000000713675 in code_store::a_store::get_thread_pimpl(std::shared_ptr<code_store::a_store::a_thread_impl>&, bool*) ()
(gdb) where
#0  0x0000000000713675 in code_store::a_store::get_thread_pimpl(std::shared_ptr<code_store::a_store::a_thread_impl>&, bool*) ()
#1  0x0000000000720e99 in code_store::a_store::get_code_items(code_store::a_code_item_query const&, std::shared_ptr<code_store::full_code_items_set>&) ()
#2  0x00000000006e2ce8 in browse_engine::query_goto_def(vscode::file_uri const&, position const&, std::vector<code_store::a_full_code_item_record, std::allocator<code_store::a_full_code_item_record> >&, std::set<std::string, std::less<std::string>, std::allocator<std::string> >&, bool, browse_engine::goto_mode, std::unique_ptr<std::string const, std::default_delete<std::string const> >, std::unique_ptr<intellisense_symbol const, std::default_delete<intellisense_symbol const> >, bool) ()
#3  0x00000000005f7bf2 in goto_definition::handle_goto_definition(vscode::TextDocumentPositionParams, browse_engine::goto_mode, bool, bool) ()
#4  0x00000000005ef841 in vscode::documentation_comment::from(vscode::documentation_comment::position const&, std::shared_ptr<goto_definition>) ()
#5  0x000000000060db72 in intellisense_client::get_quick_info(char const*, int, int, int, std::shared_ptr<goto_definition>) ()
#6  0x00000000005fd880 in hover::handle_hover(vscode::TextDocumentPositionParams, bool, std::shared_ptr<goto_definition>) ()
#7  0x0000000000653e6c in vscode::message_handler::textDocument_hover(vscode::TextDocumentPositionParams) ()
#8  0x00000000006440c2 in vscode::message_handler::dispatch(vscode::vscode_client_message&&, vscode::vscode_server_message&, vscode::message_handler::msg_proc_thread_token) ()
#9  0x0000000000641e51 in vscode::message_handler::handle_message(vscode::vscode_client_message&&, vscode::message_handler::msg_proc_thread_token)
    ()
#10 0x000000000067adf0 in std::thread::_State_impl<std::thread::_Invoker<std::tuple<vscode::message_handler::main_loop()::$_1> > >::_M_run() ()
#11 0x0000000000c6b04f in execute_native_thread_routine ()
#12 0x00007fa911cb5a92 in start_thread () from /usr/lib/libpthread.so.0
#13 0x00007fa911be3cd3 in clone () from /usr/lib/libc.so.6
```

Where did it fail? 5 lines from the bottom of this disassembly at the "=>":

```
(gdb) disassemble
Dump of assembler code for function _ZN10code_store7a_store16get_thread_pimplERSt10shared_ptrINS0_13a_thread_implEEPb:
   0x0000000000713540 <+0>:	push   %r15
   0x0000000000713542 <+2>:	push   %r14
   0x0000000000713544 <+4>:	push   %r13
   0x0000000000713546 <+6>:	push   %r12
   0x0000000000713548 <+8>:	push   %rbx
   0x0000000000713549 <+9>:	sub    $0x10,%rsp
   0x000000000071354d <+13>:	mov    %rdx,%r15
   0x0000000000713550 <+16>:	mov    %rsi,%r13
   0x0000000000713553 <+19>:	mov    %rdi,%r14
   0x0000000000713556 <+22>:	mov    0xa45c0c(%rip),%al        # 0x1159168 <_ZGVZN10code_store7a_store16get_thread_pimplERSt10shared_ptrINS0_13a_thread_implEEPbE13s_impls_cache>
   0x000000000071355c <+28>:	test   %al,%al
   0x000000000071355e <+30>:	je     0x713761 <_ZN10code_store7a_store16get_thread_pimplERSt10shared_ptrINS0_13a_thread_implEEPb+545>
   0x0000000000713564 <+36>:	mov    $0x1159164,%edi
   0x0000000000713569 <+41>:	callq  0x732510 <_ZNK9microsoft2vc14thread_local_tIN10code_store14an_impls_cacheEE3getEv>
   0x000000000071356e <+46>:	mov    %rax,%r12
   0x0000000000713571 <+49>:	mov    %rsp,%rdi
   0x0000000000713574 <+52>:	mov    %r12,%rsi
   0x0000000000713577 <+55>:	mov    %r14,%rdx
   0x000000000071357a <+58>:	callq  0x7325c0 <_ZN10code_store14an_impls_cache4findEPNS_7a_storeE>
   0x000000000071357f <+63>:	movaps (%rsp),%xmm0
   0x0000000000713583 <+67>:	xorps  %xmm1,%xmm1
   0x0000000000713586 <+70>:	movaps %xmm1,(%rsp)
   0x000000000071358a <+74>:	mov    0x8(%r13),%rbx
   0x000000000071358e <+78>:	movups %xmm0,0x0(%r13)
   0x0000000000713593 <+83>:	test   %rbx,%rbx
   0x0000000000713596 <+86>:	je     0x7135fc <_ZN10code_store7a_store16get_thread_pimplERSt10shared_ptrINS0_13a_thread_implEEPb+188>
   0x0000000000713598 <+88>:	mov    $0x57e700,%eax
   0x000000000071359d <+93>:	test   %rax,%rax
   0x00000000007135a0 <+96>:	je     0x7135b3 <_ZN10code_store7a_store16get_thread_pimplERSt10shared_ptrINS0_13a_thread_implEEPb+115>
   0x00000000007135a2 <+98>:	mov    $0xffffffff,%eax
   0x00000000007135a7 <+103>:	lock xadd %eax,0x8(%rbx)
   0x00000000007135ac <+108>:	cmp    $0x1,%eax
   0x00000000007135af <+111>:	je     0x7135c1 <_ZN10code_store7a_store16get_thread_pimplERSt10shared_ptrINS0_13a_thread_implEEPb+129>
   0x00000000007135b1 <+113>:	jmp    0x7135fc <_ZN10code_store7a_store16get_thread_pimplERSt10shared_ptrINS0_13a_thread_implEEPb+188>
   0x00000000007135b3 <+115>:	mov    0x8(%rbx),%eax
   0x00000000007135b6 <+118>:	lea    -0x1(%rax),%ecx
   0x00000000007135b9 <+121>:	mov    %ecx,0x8(%rbx)
   0x00000000007135bc <+124>:	cmp    $0x1,%eax
   0x00000000007135bf <+127>:	jne    0x7135fc <_ZN10code_store7a_store16get_thread_pimplERSt10shared_ptrINS0_13a_thread_implEEPb+188>
   0x00000000007135c1 <+129>:	mov    (%rbx),%rax
   0x00000000007135c4 <+132>:	mov    %rbx,%rdi
   0x00000000007135c7 <+135>:	callq  *0x10(%rax)
   0x00000000007135ca <+138>:	mov    $0x57e700,%eax
   0x00000000007135cf <+143>:	test   %rax,%rax
   0x00000000007135d2 <+146>:	je     0x7135e5 <_ZN10code_store7a_store16get_thread_pimplERSt10shared_ptrINS0_13a_thread_implEEPb+165>
   0x00000000007135d4 <+148>:	mov    $0xffffffff,%eax
   0x00000000007135d9 <+153>:	lock xadd %eax,0xc(%rbx)
   0x00000000007135de <+158>:	cmp    $0x1,%eax
   0x00000000007135e1 <+161>:	je     0x7135f3 <_ZN10code_store7a_store16get_thread_pimplERSt10shared_ptrINS0_13a_thread_implEEPb+179>
--Type <RET> for more, q to quit, c to continue without paging--
   0x00000000007135e3 <+163>:	jmp    0x7135fc <_ZN10code_store7a_store16get_thread_pimplERSt10shared_ptrINS0_13a_thread_implEEPb+188>
   0x00000000007135e5 <+165>:	mov    0xc(%rbx),%eax
   0x00000000007135e8 <+168>:	lea    -0x1(%rax),%ecx
   0x00000000007135eb <+171>:	mov    %ecx,0xc(%rbx)
   0x00000000007135ee <+174>:	cmp    $0x1,%eax
   0x00000000007135f1 <+177>:	jne    0x7135fc <_ZN10code_store7a_store16get_thread_pimplERSt10shared_ptrINS0_13a_thread_implEEPb+188>
   0x00000000007135f3 <+179>:	mov    (%rbx),%rax
   0x00000000007135f6 <+182>:	mov    %rbx,%rdi
   0x00000000007135f9 <+185>:	callq  *0x18(%rax)
   0x00000000007135fc <+188>:	mov    0x8(%rsp),%rbx
   0x0000000000713601 <+193>:	test   %rbx,%rbx
   0x0000000000713604 <+196>:	je     0x71366a <_ZN10code_store7a_store16get_thread_pimplERSt10shared_ptrINS0_13a_thread_implEEPb+298>
   0x0000000000713606 <+198>:	mov    $0x57e700,%eax
   0x000000000071360b <+203>:	test   %rax,%rax
   0x000000000071360e <+206>:	je     0x713621 <_ZN10code_store7a_store16get_thread_pimplERSt10shared_ptrINS0_13a_thread_implEEPb+225>
   0x0000000000713610 <+208>:	mov    $0xffffffff,%eax
   0x0000000000713615 <+213>:	lock xadd %eax,0x8(%rbx)
   0x000000000071361a <+218>:	cmp    $0x1,%eax
   0x000000000071361d <+221>:	je     0x71362f <_ZN10code_store7a_store16get_thread_pimplERSt10shared_ptrINS0_13a_thread_implEEPb+239>
   0x000000000071361f <+223>:	jmp    0x71366a <_ZN10code_store7a_store16get_thread_pimplERSt10shared_ptrINS0_13a_thread_implEEPb+298>
   0x0000000000713621 <+225>:	mov    0x8(%rbx),%eax
   0x0000000000713624 <+228>:	lea    -0x1(%rax),%ecx
   0x0000000000713627 <+231>:	mov    %ecx,0x8(%rbx)
   0x000000000071362a <+234>:	cmp    $0x1,%eax
   0x000000000071362d <+237>:	jne    0x71366a <_ZN10code_store7a_store16get_thread_pimplERSt10shared_ptrINS0_13a_thread_implEEPb+298>
   0x000000000071362f <+239>:	mov    (%rbx),%rax
   0x0000000000713632 <+242>:	mov    %rbx,%rdi
   0x0000000000713635 <+245>:	callq  *0x10(%rax)
   0x0000000000713638 <+248>:	mov    $0x57e700,%eax
   0x000000000071363d <+253>:	test   %rax,%rax
   0x0000000000713640 <+256>:	je     0x713653 <_ZN10code_store7a_store16get_thread_pimplERSt10shared_ptrINS0_13a_thread_implEEPb+275>
   0x0000000000713642 <+258>:	mov    $0xffffffff,%eax
   0x0000000000713647 <+263>:	lock xadd %eax,0xc(%rbx)
   0x000000000071364c <+268>:	cmp    $0x1,%eax
   0x000000000071364f <+271>:	je     0x713661 <_ZN10code_store7a_store16get_thread_pimplERSt10shared_ptrINS0_13a_thread_implEEPb+289>
   0x0000000000713651 <+273>:	jmp    0x71366a <_ZN10code_store7a_store16get_thread_pimplERSt10shared_ptrINS0_13a_thread_implEEPb+298>
   0x0000000000713653 <+275>:	mov    0xc(%rbx),%eax
   0x0000000000713656 <+278>:	lea    -0x1(%rax),%ecx
   0x0000000000713659 <+281>:	mov    %ecx,0xc(%rbx)
   0x000000000071365c <+284>:	cmp    $0x1,%eax
   0x000000000071365f <+287>:	jne    0x71366a <_ZN10code_store7a_store16get_thread_pimplERSt10shared_ptrINS0_13a_thread_implEEPb+298>
   0x0000000000713661 <+289>:	mov    (%rbx),%rax
   0x0000000000713664 <+292>:	mov    %rbx,%rdi
   0x0000000000713667 <+295>:	callq  *0x18(%rax)
   0x000000000071366a <+298>:	cmpq   $0x0,0x0(%r13)
   0x000000000071366f <+303>:	jne    0x713732 <_ZN10code_store7a_store16get_thread_pimplERSt10shared_ptrINS0_13a_thread_implEEPb+498>
=> 0x0000000000713675 <+309>:	mov    0x39(%r14),%al
   0x0000000000713679 <+313>:	test   $0x1,%al
   0x000000000071367b <+315>:	jne    0x71372b <_ZN10code_store7a_store16get_thread_pimplERSt10shared_ptrINS0_13a_thread_implEEPb+491>
   0x0000000000713681 <+321>:	test   %r15,%r15
   0x0000000000713684 <+324>:	je     0x71368b <_ZN10code_store7a_store16get_thread_pimplERSt10shared_ptrINS0_13a_thread_implEEPb+331>
```

Why did that line fail? Null Pointer apparently

```
(gdb) info reg
rax            0x7fa8cbffd9c0      140362953775552
rbx            0x0                 0
rcx            0x7fa911cca080      140364124823680
rdx            0x0                 0
rsi            0x7fa8c4019860      140362819672160
rdi            0x7fa8cbffd9c0      140362953775552
rbp            0x7fa8cbffdda8      0x7fa8cbffdda8
rsp            0x7fa8cbffd9c0      0x7fa8cbffd9c0
r8             0x7fa911cca080      140364124823680
r9             0x1                 1
r10            0xfffffffffffff019  -4071
r11            0x7fa911cbcf90      140364124770192
r12            0x7fa8c4019860      140362819672160
r13            0x7fa8cbffda00      140362953775616
r14            0x0                 0
r15            0x0                 0
rip            0x713675            0x713675 <code_store::a_store::get_thread_pimpl(std::shared_ptr<code_store::a_store::a_thread_impl>&, bool*)+309>
eflags         0x10246             [ PF ZF IF RF ]
cs             0x33                51
ss             0x2b                43
ds             0x0                 0
es             0x0                 0
fs             0x0                 0
gs             0x0                 0
```
It looks like R14 was loaded from a Parameter, and that might be the "bool *" one:

```
get_thread_pimpl(std::shared_ptr<code_store::a_store::a_thread_impl>&, bool*) ()
...
   0x000000000071354d <+13>:	mov    %rdx,%r15
   0x0000000000713550 <+16>:	mov    %rsi,%r13
   0x0000000000713553 <+19>:	mov    %rdi,%r14
```

**Comment by tomatghub** [comment]
> It looks like R14 was loaded from a Parameter, and that might be the "bool *" one:

Not likely as the reference is to `0x39(%r14)`.

I think I found the cause of the problem. I had the following in "settings.json":
```
    ...
    "C_Cpp.updateChannel": "Insiders",
    "C_Cpp.default.includePath": "../build",
    ...
```
instead of:
```
    ...
    "C_Cpp.updateChannel": "Insiders",
    "C_Cpp.default.includePath": ["../build"],
    ...
```
The editor objected to the former, but only by putting a "green squiggle" under it. Whatever that bad parameter did, it looks like there wasn't any further error checking on it.

The "Settings Editor" describes this as "The value to use in a configuration..." without giving the format for the "value" or stating that it is an array. I had to use Google to get a "working example" so I could copy it.

What I really meant to do was to set "includePath", but it took a while to find where to edit that, as it doesn't show up under the "Settings" menu where I expected to find it.
